plat/hisilicon/hikey/hikey_security.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2015, Linaro Ltd and Contributors. All rights reserved.
+ * Copyright (c) 2015, Hisilicon Ltd and Contributors. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * Neither the name of ARM nor the names of its contributors may be used
+ * to endorse or promote products derived from this software without specific
+ * prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <debug.h>
+#include <stdint.h>
+#include <strings.h>
+#include <platform_def.h>
+
+#define PORTNUM_MAX		5
+
+#define MDDRC_SECURITY_BASE	0xF7121000
+
+struct int_en_reg {
+	unsigned in_en:1;
+	unsigned reserved:31;
+};
+
+struct rgn_map_reg {
+	unsigned rgn_base_addr:24;
+	unsigned rgn_size:6;
+	unsigned reserved:1;
+	unsigned rgn_en:1;
+};
+
+struct rgn_attr_reg {
+	unsigned sp:4;
+	unsigned security_inv:1;
+	unsigned reserved_0:3;
+	unsigned mid_en:1;
+	unsigned mid_inv:1;
+	unsigned reserved_1:6;
+	unsigned rgn_en:1;
+	unsigned subrgn_disable:16;
+};
+
+static volatile struct int_en_reg *get_int_en_reg(uint32_t base)
+{
+	uint64_t addr = base + 0x20;
+	return (struct int_en_reg *)addr;
+}
+
+static volatile struct rgn_map_reg *get_rgn_map_reg(uint32_t base, int region, int port)
+{
+	uint64_t addr = base + 0x100 + 0x10 * region + 0x400 * port;
+	return (struct rgn_map_reg *)addr;
+}
+
+static volatile struct rgn_attr_reg *get_rgn_attr_reg(uint32_t base, int region,
+					     int port)
+{
+	uint64_t addr = base + 0x104 + 0x10 * region + 0x400 * port;
+	return (struct rgn_attr_reg *)addr;
+}
```

Comment:
Couple of possible changes for lines 64-81:
1. Define offsets rather than magic numbers.
2. Instead of functions, you could define macros.
These are not blocking from my point of view.

Reply:
These are code 'excerpts' from the vendor so we don't know what to name the defines/offsets.

Reply:
@vchong 
Is the TRM from the vendor publicly available? 
Perhaps:
- offset in get_int_en_reg could be ENABLE_REG_OFFSET.
- in get_rgn_map_reg REGION_MAP_REG_OFFSET, REGION_MAP_REGION_FIELD_SIZE, REGION_MAP_PORT_FIELD_SIZE
- in get_rgn_attr_reg REGION_ATTR_REG_OFFSET, REGION_ATTR_REGION_FIELD_SIZE, REGION_ATTR_PORT_FIELD_SIZE

But the TRM may give better 'inspiration' for the names :)

Reply:
@davidcunado-arm The TRM is not publicly available, so we'll use your suggestion. Thanks!
@petegriffin Can you please make the above recommended changes as well in v2? Thanks!

Reply:
@vchong @davidcunado-arm  Hi, I can update to macros if you want. Although I already sent the v2 patch with the other changes, so it would require another update, compile, test cycle.

Reply:
@davidcunado-arm Are you ok without updates here? You'd mentioned previously that they're not blocking.

Reply:
@vchong 
Is this part of the existing DDR IP?

Reply:
@davidcunado-arm Yes. We'd kept the patch in the https://github.com/96boards-hikey/arm-trusted-firmware fork [1][2] up to this point.

[1] https://github.com/96boards-hikey/arm-trusted-firmware/commit/38fc0955a31bd822ab8917a0d35ca3b8683d104c
[2] https://github.com/96boards-hikey/arm-trusted-firmware/commit/fb1158a365e2bf5bba638cde950678fddf67fe60

Reply:
@vchong 
Ok - all good with me them. Will progress.